arrow-core-test/src/main/kotlin/arrow/core/test/laws/FxLaws.kt

```diff
@@ -0,0 +1,166 @@
+package arrow.core.test.laws
+
+import arrow.Kind
+import arrow.core.EagerBind
+import arrow.core.test.concurrency.SideEffect
+import arrow.core.test.generators.throwable
+import arrow.typeclasses.suspended.BindSyntax
+import io.kotlintest.fail
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+private typealias EagerFxBlock<F, A> = (suspend EagerBind<F>.() -> A) -> Kind<F, A>
+private typealias SuspendFxBlock<F, A> = suspend (suspend BindSyntax<F>.() -> A) -> Kind<F, A>
+
+object FxLaws {
+
+  fun <F, A> laws(
+    pureGen: Gen<Kind<F, A>>, // TODO cannot specify or filter a pure generator, so we need to require an additional one
+    G: Gen<Kind<F, A>>,
+    fxBlock: EagerFxBlock<F, A>,
+    sfxBlock: SuspendFxBlock<F, A>
+  ): List<Law> = listOf(
+    Law("non-suspended fx is lazy") { nonSuspendedIsLazy(G, fxBlock) },
+    Law("suspended fx is lazy") { suspendedIsLazy(G, sfxBlock) },
+    Law("non-suspended fx can bind immediate values") { nonSuspendedCanBindImmediate(G, fxBlock) },
+    Law("non-suspended fx can bind immediate exceptions") { nonSuspendedCanBindImmediateException(pureGen, fxBlock) },
+    Law("suspended fx can bind immediate values") { suspendedCanBindImmediateValues(G, sfxBlock) },
+    Law("suspended fx can bind suspended values") { suspendedCanBindSuspendedValues(G, sfxBlock) },
+    Law("suspended fx can bind immediate exceptions") { suspendedCanBindImmediateExceptions(pureGen, sfxBlock) },
+    Law("suspended fx can bind suspended exceptions") { suspendedCanBindSuspendedExceptions(pureGen, sfxBlock) }
+  )
+
+  private suspend fun <F, A> nonSuspendedIsLazy(G: Gen<Kind<F, A>>, fxBlock: EagerFxBlock<F, A>) {
+    forAll(G) { f: Kind<F, A> ->
+      val effect = SideEffect()
+      fxBlock {
+        effect.increment()
+        f.bind()
+      }
+
+      effect.counter == 0
```

Comment:
This should be `1` for an eager block, no?

Reply:
I removed the lazy tests, basically because the suspended fx block is lazy by default, and having the non-suspend one to be lazy will cause issues with the other tests, so better leave like this I think